chain/messagepool.go

```diff
@@ -328,15 +374,18 @@ func (mp *MessagePool) Remove(from address.Address, nonce uint64) {
 	delete(mset.msgs, nonce)
 
 	if len(mset.msgs) == 0 {
-		// FIXME: This is racy
-		//delete(mp.pending, from)
+		delete(mp.pending, from)
 	} else {
 		var max uint64
 		for nonce := range mset.msgs {
 			if max < nonce {
 				max = nonce
 			}
 		}
+		if nonce > max {
+			max = nonce // we could have not seen the removed message before
+		}
+
```

Comment:
One note: we should remove messages with lower nonce than than `nonce` from function input.